common/xPowerManager.h

```diff
@@ -0,0 +1,769 @@
+
+    // #define DIAGNOSE_POWER_PORT
+    // Timeouts
+    #define PWRMAN_TIMEOUTRES   10000       // timeout resolution, in micros
+    #define PWRMAN_MINTIMEOUT   20          // minimum timeout, in millis
+    #define PWRMAN_DEFTIMEOUT   1000        // default timeout: 1 second
+    #define PWRMAN_AMPTIMEOUT   50          // audio amplifier timeout 
+    #define PWRMAN_CPUTIMEOUT   60000       // CPU deep sleep timeout
+    #define PWRMAN_SDMOUNTTIMEOUT 5000      // SD mount timeout - allow longer for pre-loop initializations
+    #define PWRMAN_STARTON  pwr4_CPU | pwr4_SD  // Binary map of PDType flags of domains that will turn ON at startup
+
+    // Stop entry type 
+    #define PWR_STOPENTRY_WFI               ((uint8_t)0x01)       //Wait For Interruption instruction to enter Stop mode
+    #define PWR_STOPENTRY_WFE               ((uint8_t)0x02)       //Wait For Event instruction to enter Stop mode
+
+    // Types of power domains 
+    #define PDType_base uint8_t
+    enum PDType : PDType_base { 
+        pwr4_none       = 0,
+        pwr4_CPU        = 0b00000001,    
+        pwr4_SD         = 0b00000010,
+        pwr4_Booster    = 0b00000100,
+        pwr4_Amplif     = 0b00001000,
+        pwr4_Pixel      = 0b00010000,
+        pwr4_Charger    = 0b00100000,
+    };        
+
+    // Wake-up sources
+    enum WKSource {
+        wakeUp_none = 0,
+        wakeUp_button = 1, // wake up from button
+        wakeUp_serial = 2, // wake up from serial
+    };
+
+    class xPowerDomain;        
+    class xPowerSubscriber;
+
+    class xPowerManager : public Looper, CommandParser    
+    {   friend class xPowerDomain;
+        friend class xPowerSubscriber;
+        
+    private:
+        xPowerDomain* domains;           // linked list of domains
+        xPowerSubscriber* subscribers;   // linked list of subscribers
+        static WKSource wakeUpSource;   // need this static because we need static ISRs to access it
+        PDType_base powerState;         // binary map of all active 'PDType' flags
+        uint32_t lastLoopTime;          // time keeper
+        
+        void Setup() override;          // delayed implementation, see end of file                          
+        void Loop() override;           
+
+    public:
+        xPowerManager() : Looper() {    // Looper(PWRMAN_TIMEOUTRES)
+            powerState = 0;
+            domains = NULL;
+            lastLoopTime = 0;
+        }
+        const char* name() override { return "xPowerManager"; }
+
+        // Turn ON power domains 
+        bool Activate(PDType_base startUpDomains = PWRMAN_STARTON);            // delayed implementation, see end of file.
+        bool Parse(const char* cmd, const char* arg) override;   // delayed implementation, see end of file.    
+
+    private:
+        static void initWakeups()
+        {
+            // set btn and VUSB pin for for wakeup 
+            stm32l4_exti_notify(&stm32l4_exti, g_APinDescription[powerButtonPin].pin,
```

Comment:
It doesn't really matter because with no button the processor will not go in deep sleep - it is protected by prop.HoldPower(). The assumption here is that if at least one button in installed, it will be tied at powerButtonPin.

Reply:
What I mean is that someone could be driving a LED from that pin, or use it as a touch button, or something completely different. We can't assume that it's a power button unless there is a Button class tied to it.